BlueDBOnDisk/src/main/java/org/bluedb/disk/backup/BackupManager.java

```diff
@@ -58,39 +82,124 @@ private void copyMetaData(ReadWriteCollectionOnDisk<?> collection, Path tempFold
-			try (BlueReadLock<Path> lock = segment.getReadLockFor(groupingNumber)) {
-				Path src = lock.getKey();
-				Path dst = translatePath(dbPath, tempFolder, src);
-				FileUtils.copyFileWithoutLock(src, dst);  // already have read lock on src, shouldn't need write lock on dst
+			
+			if(isTimeBased && !includedTimeRange.isMaxRange()) {
+				copyDataFileAfterFilteringBasedOnTime(segment, tempFolder, includedTimeRange, groupingNumber);
+			} else {
+				copyDataFileStraightOver(segment, tempFolder, groupingNumber);
 			}
 		}
 	}
 
-	private void copyChanges(ReadWriteCollectionOnDisk<?> collection, long backupStartTime, long backupEndTime, Path tempFolder) throws BlueDbException {
+	private void copyDataFileAfterFilteringBasedOnTime(ReadWriteSegment<?> segment, Path tempFolder, Range includedTimeRange, long groupingNumber) throws BlueDbException {
+		Path src = segment.getPathFor(groupingNumber);
+		Path dst = translatePath(dbPath, tempFolder, src);
+		FileUtils.ensureDirectoryExists(dst.toFile());
+		boolean isFileEmpty = true;
+		
+		try(BlueObjectOutput<BlueEntity<?>> output = BlueObjectOutput.createWithoutLockOrSerializer(dst)) {
+			try(BlueObjectInput<?> fileInputStream = segment.getObjectInputFor(groupingNumber)) {
+				while(fileInputStream.hasNext()) {
+					BlueEntity<?> next = (BlueEntity<?>) fileInputStream.next();
+					if(next.getKey().isInRange(includedTimeRange.getStart(), includedTimeRange.getEnd())) {
+						output.writeBytes(fileInputStream.getLastBytes());
+						isFileEmpty = false;
+					}
+				}
+			}
+		}
+		
+		if(isFileEmpty) {
+			FileUtils.deleteIfExistsWithoutLock(dst);
+		}
+	}
+
+	private void copyDataFileStraightOver(ReadWriteSegment<?> segment, Path tempFolder, long groupingNumber) throws BlueDbException {
+		try (BlueReadLock<Path> lock = segment.getReadLockFor(groupingNumber)) {
+			Path src = lock.getKey();
+			Path dst = translatePath(dbPath, tempFolder, src);
+			FileUtils.copyFileWithoutLock(src, dst);  // already have read lock on src, shouldn't need write lock on dst
+		}
+	}
+
+	private void copyChanges(ReadWriteCollectionOnDisk<?> collection, Range includedTimeRange, Range includedChangeTimes, Path tempFolder) throws BlueDbException {
 		RecoveryManager<?> recoveryManager = collection.getRecoveryManager();
-		List<File> changesToCopy = recoveryManager.getChangeHistory(backupStartTime, backupEndTime);
+		List<File> changesToCopy = recoveryManager.getChangeHistory(includedChangeTimes.getStart(), includedChangeTimes.getEnd());
 		Path historyFolderPath = recoveryManager.getHistoryFolder();
 		Path destinationFolderPath = translatePath(dbPath, tempFolder, historyFolderPath);
 		destinationFolderPath.toFile().mkdirs();
 		for (File file: changesToCopy) {
 			Path destinationPath = Paths.get(destinationFolderPath.toString(), file.getName());
-			FileUtils.copyFileWithoutLock(file.toPath(), destinationPath);
+			if(!includedTimeRange.isMaxRange() && collection.isTimeBased()) {

```

Comment:
This is the second time I've seen this exact combination of conditionals in an if statement. It might be better to move them to a well named method.

Reply:
Good call, I'll do that